Implement gesture typing

X-yl opened the pull request:
Closes #139 

Most of the actual gesture classifying code was ~stolen from~ inspired by EtienneDesticourt's amazing work on the ASK repo: https://github.com/AnySoftKeyboard/AnySoftKeyboard/pull/1870

I found that the constants he obtained worked fine for me, however in the future we might like to get more data. 

The other issue is that currently it doesn't use the Dictionary API to load words. It includes a duplicate copy of a part of the word-freq dataset in JSON format. This leads to an increase in APK size of about 250KB. 

@patrickgold you should probably patch this line:
https://github.com/X-yl/florisboard/blob/cf1c18aa7021dbbb81e92daff87e186c6690b550/app/src/main/java/dev/patrickgold/florisboard/ime/text/TextInputManager.kt#L336
to support dictionaries.

So that's about it. Let me know if you catch any bugs and I'll squash them

(hype for v0.4.0 :smile:)

tsiflimagas commented:
I never systematically used glide typing, but wow, this already works really smooth! The only problem I could find is that keyboard crashes when using glide typing on non latin layouts (is this to be expected because there are no dictionaries for other languages than english?). After the crash though the feature is broken, you can glide your finger and see the trail but nothing happens. Here's the log if there's anything to investigate in this https://del.dog/ugnirexogy
Great job!

patrickgold commented:
Thanks for your PR! I will review and merge your changes after the `subtype-specific-layouts` branch is merged in, which hopefully should be tomorrow. On a quick look at the source code and mock-up from you though it looks really good :)

ildar commented:
I just tested. You sure it's ready to merge?
1. it accepts only one word out of three (no suggestions, no reaction)
2. it completely freezes after a few attempts

X-yl commented:
> I never systematically used glide typing, but wow, this already works really smooth! The only problem I could find is that keyboard crashes when using glide typing on non latin layouts (is this to be expected because there are no dictionaries for other languages than english?). After the crash though the feature is broken, you can glide your finger and see the trail but nothing happens. Here's the log if there's anything to investigate in this https://del.dog/ugnirexogy
> Great job!

Thanks for spotting this! It's definitely not supposed to crash. I will be pushing out a fix for this shortly (I'm making the glide trail themable + some other fixes, and then I'll push it all out together)

> edit: I'm also wondering if a fade-out of the trail after a glide session is completed could be implemented. And if it's possible, would it fit well or would it cause any issues? Gboard has this animation and it looks nice (though it maybe could be considered too long).

I've added this as well, stay tuned ;)

@ildar could you share some more information (perhaps a video)? What device and version of android are you trying this on? I haven't been able to replicate these issues.

Is it the first two glide words it ignores? And then it works fine? If so it could just be that the classifier is still loading asynchronously. On my device it takes about half a second, so possibly it's just taking a little longer on yours.

I don't really understand the freezing though, that shouldn't be possible, all the expensive work happens on a different thread, so it should never freeze the UI.

Glitchy-Tozier commented:
Wooow if that works you've done an amazing job @X-yl  <3

X-yl commented:
Hang on, don't merge yet, my multithreading code is really awful and needs some work. Just give me a little longer to iron out some issues.

tsiflimagas commented:
You even made fade duration configurable! It looks great!

D3SOX commented:
Not sure if it is related but I got this error using your `gesture-typing` branch:

```
~~~ 1617393659572.stacktrace ~~~

kotlin.UninitializedPropertyAccessException: lateinit property asyncExecutor has not been initialized
	at dev.patrickgold.florisboard.ime.core.FlorisBoard.getAsyncExecutor(FlorisBoard.kt:135)
	at dev.patrickgold.florisboard.ime.clip.provider.FlorisContentProvider.initIfNotAlready(FlorisContentProvider.kt:40)
	at dev.patrickgold.florisboard.ime.clip.FlorisClipboardManager$initialize$1.run(FlorisClipboardManager.kt:287)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
	at java.lang.Thread.run(Thread.java:923)

```

X-yl commented:
> Not sure if it is related but I got this error using your `gesture-typing` branch:
> 
> ```
> ~~~ 1617393659572.stacktrace ~~~
> 
> kotlin.UninitializedPropertyAccessException: lateinit property asyncExecutor has not been initialized
> 	at dev.patrickgold.florisboard.ime.core.FlorisBoard.getAsyncExecutor(FlorisBoard.kt:135)
> 	at dev.patrickgold.florisboard.ime.clip.provider.FlorisContentProvider.initIfNotAlready(FlorisContentProvider.kt:40)
> 	at dev.patrickgold.florisboard.ime.clip.FlorisClipboardManager$initialize$1.run(FlorisClipboardManager.kt:287)
> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
> 	at java.lang.Thread.run(Thread.java:923)
> ```

Uh that's related to the clipboard manager. Could you tell me when and how it crashed? I'll resolve the issue if I can track it down.

> Other that it looks very nice but it's not as accurate as Gboard. It sometimes writes `Heli` instead of `Hello`. Maybe the suggestions could provide a value how much a word is generally used so that the glide typing can consider that

This is sorta intended. You need to do a little loop over a letter that repeats. This allows the algorithm to differentiate between words like tool and toll, feel and fell etc. The algorithm already takes into account word frequency. 

But actually, I can see what you mean and why this is slightly weird. I will see if it's possible to have multiple correct gestures so that it will prefer hello over heli but still allow to differentiate tool and toll.

X-yl commented:
So I've fixed the issue where gesture typing wasn't triggering and resolved the merge conflicts. I have also added in the multiple correct gestures thing, so now swiping h-e-l-o suggests hello, but you can still do a little loop if you want to indicate a double letter. I think it's all good for you to review now @patrickgold :+1:

patrickgold commented:
@X-yl Thanks for your amazing work on glide typing, I really appreciate it! I've tested it out locally on my testing devices and it works surprisingly well and smoothly. The generated words are mostly the words I tried to type in, with a few exceptions but that is fully acceptable. I especially like that you also gave some good customization preferences, allow to set the glide color and that you also modified all existing themes to have matching colors! <3

Now to the code review:  
In general, your code looks really clean and is written very understandable, well done! I've found issues though, some where I am unsure if they are a byproduct of the merge conflict resolving, some where I generally have found something to be problematic and some where I am basically nitpicking about the syntax xD. I will leave a single-comment review on each code part where I think that it needs some fixing before I can merge in your changes.

Finally I also want to thank all other contributors who helped in chasing away initial bugs for glide typing, I really appreciate your help!

patrickgold commented on app/src/main/java/dev/patrickgold/florisboard/ime/text/TextInputManager.kt:
```diff
@@ -323,21 +325,31 @@ class TextInputManager private constructor() : CoroutineScope by MainScope(), In
+                        ).await()
+                        keyboardView.updateVisibility()
+                    }
                 }
```
The whole `for (keyboardMode in KeyboardMode.values())` for loop should not be within this if statement, as then the layouts upon a subtype change in the preferences only get updated if glide typing is enabled. Layout re-generation must be done though regardless of glide typing, so it should be placed either above or below the if statement.

patrickgold replied:
Whoops, I accidentally put the comment on the wrong side here, I meant of course L336-346 of your TextInputManager file.

patrickgold commented on app/src/main/java/dev/patrickgold/florisboard/ime/text/layout/Layout.kt:
```diff
@@ -61,14 +61,14 @@ data class LayoutMetaOnly(
 
 typealias ComputedLayoutArrangement = MutableList<MutableList<FlorisKeyData>>
-data class ComputedLayout(
+data class LayoutData(
```
Here I am unsure what exactly happened to the `ComputedLayout` class name, which has been introduced by me in the `subtype-specific-layouts` branch but then been renamed by you to `LayoutData`. I've chosen `ComputedLayout`, because it clarifies that this layout is dynamically computed, whereas `LayoutData` (or `Layout` since the `subtype-specific-layouts` branch) is the original layout from the json file.

X-yl replied:
What I did was I renamed ComputedLayout back to ComputedLayoutData (so that I could then easily rename it back to ComputedLayout to fix all the code which had it as ComputedLayoutData). It seems then that I somehow forgot what it was called (how did I do that??) and named it LayoutData. My bad, will fix.

patrickgold commented on app/src/main/java/dev/patrickgold/florisboard/settings/fragments/GesturesFragment.kt:
```diff
@@ -16,12 +16,38 @@
 import dev.patrickgold.florisboard.R
 
-class GesturesFragment : PreferenceFragmentCompat() {
+class GesturesFragment : PreferenceFragmentCompat(), SharedPreferences.OnSharedPreferenceChangeListener {
+    private lateinit var sharedPreferences: SharedPreferences
     override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
         addPreferencesFromResource(R.xml.prefs_gestures)
+        sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context)
+        val preference: Preference? = findPreference("glide__show_trail")
+        preference?.isEnabled = sharedPreferences.getBoolean("glide__enabled", false)
     }
+
+    override fun onResume() {
+        super.onResume()
+        sharedPreferences.registerOnSharedPreferenceChangeListener(this);
+    }
+
+    override fun onPause()
+    {
+        super.onPause()
+        sharedPreferences.unregisterOnSharedPreferenceChangeListener(this);
+    }
+
+    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, key: String) {
+        if (key == "glide__enabled"){
+            val preference: Preference? = findPreference("glide__show_trail")
+            preference?.isEnabled = sharedPreferences.getBoolean("glide__enabled", false)
+        }
+    }
+
```
Here I am unsure why you are building a custom logic for dynamically enabling/disabling a preference, as the `app:dependency="glide__enabled"` attribute on the `glide__show_trail`-preference would basically do the same work for you and it also disables all sub-dependencies automatically.

X-yl replied:
Ah this one has a good explanation lol. The issue is that some of the preferences need to be disabled (the preferences for general gestures) whereas some need to be enabled (the glide trail options). From what I can tell though, `app:dependency` doesn't allow this, it only allows one or the other through the `app:disableDependentsState` option. That's why I had to implement the logic myself.

patrickgold replied:
Oh I can remember now, I used the `app:disableDependentsState` for disabling the swipe left/right/up/down gesture prefs. That's one of the big main restrictions of the dependency attribute sadly. Good, I will mark this one as resolved then :)

patrickgold commented on app/src/main/java/dev/patrickgold/florisboard/ime/text/keyboard/KeyboardView.kt:
```diff
@@ -107,11 +126,41 @@ class KeyboardView : FlexboxLayout, FlorisBoard.EventListener, SwipeGesture.List
     }
 
+    private fun initGestureClassifier() {
+        if (this.isSmartbarKeyboardView ||
+            this.computedLayout?.mode != KeyboardMode.CHARACTERS
+        )
+            return
```
Basically nitpicking by me, but this `return` statement at first glance doesn't look like it is in the body of the above if statement, I'd suggest wrapping it with curly braces and adding newline separators to make this more clear.

patrickgold commented on app/src/main/java/dev/patrickgold/florisboard/ime/text/keyboard/KeyboardView.kt:
```diff
@@ -394,6 +482,12 @@ class KeyboardView : FlexboxLayout, FlorisBoard.EventListener, SwipeGesture.List
+        if (theme.getAttr(Theme.Attr.GLIDE_TRAIL_COLOR).toSolidColor().color == 0){
+            this.glideTrailPaint.color = theme.getAttr(Theme.Attr.WINDOW_COLOR_PRIMARY).toSolidColor().color
+            this.glideTrailPaint.alpha = 32
+        }else {
```
Here you forgot a whitespace before the `else` :)

patrickgold commented on app/src/main/java/dev/patrickgold/florisboard/ime/text/key/KeyData.kt:
```diff
@@ -232,7 +232,11 @@ class FlorisKeyData(
     var popup: PopupSet<KeyData> = PopupSet(),
-    var shift: KeyData? = null
+    var shift: KeyData? = null,
+    var x: Float = 0f,
+    var y: Float = 0f,
+    var height: Int = 0,
+    var width: Int = 0
```
Here you are adding layout data to the `KeyData` object. While this of course works, generally the key data is designed to hold the data describing what this key is (a character, symbol and what code should be send when it is pressed), but not any layout-related data (that's what `KeyView` is responsible for). I am unsure how difficult it is to fix this now. If it is a relatively easy fix, I'd suggest that you save this information inside the `KeyView` and leave `KeyData` how it is.

X-yl replied:
Given that `ComputedLayoutData` was just a list of these `KeyData`s, I thought it would be natural to include the actual coordinates as part of the data. That and, it made it a little easier to port the ASK code, since this is how they were storing their data ;) . It shouldn't be a huge deal to move this code to `KeyView`, so I'll do that.

X-yl commented:
Alright fixed, I think. Let me know if you want me to change anything else!

patrickgold commented:
Perfect, thanks for fixing!

While playing around with glide typing for a while, I found a bug when touching the screen with 2+ input pointers: When gliding over the keyboard with the initial pointer, everything works fine. Then wile still gliding with the first pointer I tap another key with my second pointer. What happens is that the glide of the first pointer is cancelled, but the glide trail is now stuck on the screen until I start the next glide.

Other than that, I've been unable to fool your glide logic and everything works fine, so I guess I can merge your changes once that small bug is fixed.

X-yl commented:
Okay should be fixed :)

Glitchy-Tozier commented:
1. Is there even a reason to even allow multi-touch while swiping?
I can't come up with use-cases.
1. What happens if the user switches to another view while swiping? If the swiping isn't cancled, that sounds like an issue to me.

X-yl commented:
>1. Is there even a reason to even allow multi-touch while swiping?

Not really. In fact I can probably simplify the code a bit, since I originally just duplicated SwipeGesture, which does include some of the framework for detecting multiple gestures, although it doesn't use it.

>2. What happens if the user switches to another view while swiping? If the swiping isn't cancled, that sounds like an issue to me.

I did overlook this, oops. Currently if the keyboard is hidden, such as by locking the screen or interacting with something else, it doesn't remove the gesture trail.

patrickgold commented:
Thanks a lot for fixing the issue! I have tested it out and it doesn't get stuck anymore. As there are no other issues from my side anymore, I'll merge your changes in to master now. Gesture typing will be part of the v0.3.10-beta06 release and this weekend of the stable v0.3.10 release (if no critical bugs occur until then).

EtienneDesticourt commented:
Thanks for the shootout, I'm happy to see that my work was useful to the broader FOSS community.  
Did you get a chance to implement the suggested improvements that I put in the README? In my implementation the search can be a bit slow because length-pruning is in O(n). The fix is really easy (just trade compute for memory by precalculating the gesture lengths and using the appropriate data structure) but I can't work on it right now.
Did you find any improvements on your side of things that I could add to the original PR? 

Also if anybody is up for it it'd be helpful if some people would type the dataset in my PR and share the gesture coordinates logs. Having an open dataset of gestures would enable future work on gesture detection and allow FOSS keyboards to compete with commercial alternatives.

X-yl commented:
> Did you get a chance to implement the suggested improvements that I put in the README? In my implementation the search can be a bit slow because length-pruning is in O(n). The fix is really easy (just trade compute for memory by precalculating the gesture lengths and using the appropriate data structure) but I can't work on it right now.

I didn't see that in the readme actually.. I thought it was the same as the PR writeup haha.. that is probably a good improvement though, I'll definitely add it. 

> Did you find any improvements on your side of things that I could add to the original PR? 

Not really to be honest.. it worked pretty well. One issue I noticed is that the algorithm seems to prefer accuracy over word frequency -- that would probably be corrected if we adjust the parameters though. 

One thing I wanted to add was the ability for the parameters to be adjusted on the fly. I presume I need something like Welford's online algorithm, right? Stats is not my strength lol. 

> Also if anybody is up for it it'd be helpful if some people would type the dataset in my PR and share the gesture coordinates logs. Having an open dataset of gestures would enable future work on gesture detection and allow FOSS keyboards to compete with commercial alternatives.

I think this is a great idea. A good open dataset could be really beneficial to all open source keyboards. 

@patrickgold what do you think about some opt-in feature like this that allows interested users to save their glide typing data? We could maybe have a git repo (or something that's easier to anonymize) where they could upload their data so that smart people like Etienne can write even better algorithms.. and if it gets big enough, maybe we could even do proper ML. 

Thanks again for your work on this, I definitely wouldn't have been able to add this without your implementation to go off of.

khimaros commented:
for reference this is the test set @EtienneDesticourt created and mentioned above. if there is a build of FlorisBoard or way to enable traces, i'd be happy to type these out and contribute them.

```
    "If I had a choice I'd eat the salad in my fridge rather than junk food but I'm hungry",
    "I have to wear my lenses when I use my needle holder to find the spots",
    "thomas drove martin to the pharmacy to get his ambien and zoloft",
    "in recent months uganda has been retrofitting the engines of its fleet of planes to achieve super sonic flight",
    "every year jack invites nicole and janet to taiwan for the techno festival that he throws",
    "The life saving german serum hardly helps with my asthma",
    "I'm afraid the blue smoke escaped from my tablet while I was trying to debug linux",
    "philip took photos of many cool places like york and prague",
    "you won so you can redeem your prize in quebec",
    "I have to inform howard that he has weak teeth",
    "I shall teach you the tale of gordon the copper",
    "uncle murphy likes diesel trucks from nissan",
    "the rivers of nevada supply the united states with most of its coral",
    "the barrel was dirty because the hunter didn't practice good weapon maintenance",
    "always wear gloves when you commit a crime",
    "major andrew got his injury in kuwait fighting saddam",
    "soviet russia claims that the launch of the probe was a success",
    "the rhythm of samba felt like a trance to arthur after a few drinks",
    "a second tier hammer will easily damage your floors and carpet",
    "if you gotta smoke weed make sure to do it in europe",
    "I wish I were wealthy then I'd buy a villa",
    "the actor travelled abroad to study arabic",
    "the alpha eats the spoils alone",
    "cindy chose to go to the circus to clear her mind",
    "the empire built by eminem is made of many firms",
    "I assure you that's not the correct amount, I advise you count it again",
    "craig framed his comics",
    "the world atlas committee denied the existence of these tribes",
    "it's a hard knock life having to do all this manual labor",
    "the honor of your house is at stake",
    "I made those wheels from a willow tree I found in these woods",
    "julia loves to edit ebooks with emacs",
    "angela the answer is in prayer",
    "to create the new crown I came up with a cooler design that includes a lot of green and gray",
    "the scientists assess the potency of the acids",
    "boston and tampa are both boring",
    "the owner gave his token to the highest bidder who ended up paying twelve million dollars",
    "follow the gnome, his name is ronald wesley",
    "the boards have their finger on the pulse of public opinion",
    "its only arbor is in the arctic sea",
    "It's urgent that you upload this weblog to the data vault",
    "poison has a big impact on the immune system",
    "watson was focused on forcefully shoving garlic in the turkey",
    "there is little logic in wearing lose belts",
    "robin took action when faced with racial discrimination",
    "duncan draws a duck",
    "this medieval family favors incest",
    "the winds of this canyon can break a window",
    "the pilot picked the oasis to crashland",
    "pretty poetry can be short and simple",
    "before he bought his quilt, murray bought briefs",
    "the intro to the dead island video game is great",
    "the saints had a vision of christ and joseph in the chapel",
    "asus aspires to branch out into the entreprise market before the end of the fiscal year",
    "disney payed the courts to change copyright laws",
    "solar eclipses are best viewed with glasses",
    "the people of vernon see the value in verbal assault",
    "this is the fifth film I've watched instead of preparing my finals",
    "the garmin organization is flush with the funds collected for their cause of saving gay indian infants",
    "as soon as a flamethrower appeared, everybody would shoot at it, it's better than having your skin slide off your body"
```

patrickgold commented:
> @patrickgold what do you think about some opt-in feature like this that allows interested users to save their glide typing data? We could maybe have a git repo (or something that's easier to anonymize) where they could upload their data so that smart people like Etienne can write even better algorithms.. and if it gets big enough, maybe we could even do proper ML.

This definitely sounds like a great idea. As long as it is opt-in (by default everything within FlorisBoard is private) I am very open towards supporting such projects. Especially if we are able to find a format which is used across several FOSS keyboards the usefulness of such data increases a lot. Rn I am pretty occupied (I have my finals in college) but soon I will be able to focus more on FlorisBoard again, then I can definitely offer my help should it be needed :)

> for reference this is the test set @EtienneDesticourt created and mentioned above. if there is a build of FlorisBoard or way to enable traces, i'd be happy to type these out and contribute them.

Currently there's no way, I'm sorry, but experimenting with implementing such feature would definitely go in line with the above-mentioned idea.

X-yl commented:
Great! I'll work on it then. @EtienneDesticourt what data do you want me to save? I presume we'll just need to save the user gesture, with the coordinates normalized to be between zero to one, plus the word they meant to type -- so if they gesture a word and then chose a suggestion, then the suggestion they picked, or if they deleted and retyped a different word, then the word they retyped. Correct?

Glitchy-Tozier commented:
If we need this data for machine learning, we need to add something: 
1. Where the user came from before the word (= the previous two letters)
2. Where the user went after the current word (= the following two letters)

Of course those data are more difficult to track as there aren't always previous/next words, but they are *really important*! The previous/next letters can and do distort the movement towards them, after all.

janicez commented:
I'm glad i followed this link trail. While it's not perfect, i've found this glide typing algorithm mostly usable for the corpus that i generate

EtienneDesticourt commented:
@X-yl In my implementation I just saved the raw coordinates, not normalized. With enough data it's easy enough to do the normalization à posteriori. Nothing else is needed when people type a standardized set of sentences since the set of words is very small and they always come in the same sequence, you can infer what the users meant to type, which word they failed to type and retyped, etc ... from the gesture data alone. I'd imagine it'd be much simpler to log as well.
Ideally you should preserve the order of the gestures when logging. I have a program to trace the logging data, all gestures end to end, and it allows me to check that the dataset was correctly typed.

@Glitchy-Tozier That data is already included when you log a full sentence, but in my experience it's not that important, you lift your thumb between words so inter-word letter sequences don't matter as much as intra-word sequences. The dataset was made to maximize the variety of letter-trigrams (which is why the sentences are so weird) so it should give a good data on that.

khimaros commented:
@X-yl -- did you end up producing the debug build for this?

X-yl commented:
@khimaros I'm afraid not, life just caught up and I haven't had the time or motivation to work on this in quite a while. 

I'm also afraid I can't promise anything as far as working on this goes but if something sparks my interest again I might have a look at what can be done.
